### PR TITLE
Enforce true paragraphs in CKEditor

### PR DESCRIPTION
**Pull request by dtdesign**
This is an attempt to restore the behavior of “true” paragraphs in CKEditor. Pressing Enter in the editor will yield a new paragraph while Shift + Enter inserts a line break.

The change is backwards compatible by flagging newly generated, empty paragraphs with an explicit attribute. Existing empty paragraphs generated in earlier versions are detected in the output and transparently stripped. This approach does not require any changes whatsoever to both existing messages as well as any programmatic implementations that generate HTML.

### CODE REVIEW

File: wcfsetup/install/files/lib/system/html/output/node/HtmlOutputNodeNormalizer.class.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+ * Normalizes HTML generated by earlier version of WoltLab Suite.
+ *
+ * @author Alexander Ebert
+ * @copyright 2001-2023 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since 6.0
+ */
+final class HtmlOutputNodeNormalizer
+{
+    public function __construct(private readonly \DOMXPath $xpath)
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/html/output/node/HtmlOutputNodeNormalizer.class.php:17]: This parameter should be a parameter of `normalize()` and then passed along. This removes the implicit dependency on `$this->xpath` in some methods (not all need access to the xpath!) and would make objects of the normalizer reusable for normalizing multiple messages.

File: wcfsetup/install/files/lib/system/html/output/node/HtmlOutputNodeNormalizer.class.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                // We need to reduce the number of paragraphs by half, unless it
+                // is an uneven number in which case we need to remove one
+                // additional paragraph.
+                if ($offset % 2 === 1) {
+                    // 2 -> 1, 4 -> 2
+                    $numberOfParagraphsToRemove = \ceil($offset / 2);
+                } else {
+                    // 3 -> 1, 5 -> 2
+                    $numberOfParagraphsToRemove = \ceil($offset / 2) + 1;
+                }
```
[the remark this answers is not in the document]
Comment (dtdesign) [reply wcfsetup/install/files/lib/system/html/output/node/HtmlOutputNodeNormalizer.class.php:92]: No, it is correct. `$offset` refers to the number of paragraphs that follow `$candidate`. When there are two consecutive paragraphs in total that should be trimmed, then `$offset === 1`.

[the remark this answers is not in the document]
Comment (TimWolla) [reply wcfsetup/install/files/lib/system/html/output/node/HtmlOutputNodeNormalizer.class.php:92]: Ah, I see. Perhaps `$offset` should be named `$numberOfAdjacentParagraphs` then?

------------

I also must admit that I find it terribly confusing that the comment shows the number of original and the number of resulting paragraphs, but the calculation calculates neither: It calculates the difference, using the number of original paragraphs less one. In fact by using the number of original paragraphs the calculation becomes much simpler:

We need to remove half the number of paragraphs (rounded up):

```suggestion
                $totalNumberOfParagraphs = $offset + 1;
                $numberOfParagraphsToRemove = \ceil($totalNumberOfParagraphs / 2);
```

5 original paragraphs: `\ceil(5 / 2) = \ceil(2.5) = 3`, which is exactly the same as `\ceil(4 / 2) + 1`

File: ts/WoltLabSuite/Core/Component/Ckeditor/Cleanup.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      // We need to reduce the number of paragraphs by half, unless it
+      // is an uneven number in which case we need to remove one
+      // additional paragraph.
+      if (offset % 2 === 1) {
+        // 2 -> 1, 4 -> 2
+        numberOfParagraphsToRemove = Math.ceil(offset / 2);
+      } else {
+        // 3 -> 1, 5 -> 2
+        numberOfParagraphsToRemove = Math.ceil(offset / 2) + 1;
+      }
```
Comment (TimWolla) [inline ts/WoltLabSuite/Core/Component/Ckeditor/Cleanup.ts:119]: This should be unified with the PHP logic.

**Review by TimWolla — COMMENTED** [review]
LGTM other than that consistency remark. Of course I can't comment on whether the implemented functionality is the desired functionality.

**Pull request merged**